src/powerbox/tools.py

```diff
@@ -175,9 +242,227 @@ def _get_binweights(coords, weights, bins, average=True, bin_ave=True, log_bins=
     return indx, bins, sumweights
 
 
+def _spherical2cartesian(r, phi_n):
+    r"""Convert spherical coordinates to Cartesian coordinates.
+
+    Parameters
+    ----------
+    r_n : array-like
+        1D array of radii.
+    phi_n : array-like
+        2D array of azimuthal angles with shape (ndim-1, N), where N is the number of points.
+        phi_n[0,:] :math:`\in [0,2*\pi]`, and phi_n[1:,:] :math:`\in [0,\pi]`.
+
+    Returns
+    -------
+    coords : array-like
+        2D array of Cartesian coordinates with shape (ndim, N).
+
+    For more details, see https://en.wikipedia.org/wiki/N-sphere#Spherical_coordinates
+
+    """
+    if phi_n.shape[0] == 1:
+        return r * np.array([np.cos(phi_n[0]), np.sin(phi_n[0])])
+    elif phi_n.shape[0] == 2:
+        return r * np.array(
+            [
+                np.cos(phi_n[0]),
+                np.sin(phi_n[0]) * np.cos(phi_n[1]),
+                np.sin(phi_n[0]) * np.sin(phi_n[1]),
+            ]
+        )
+    else:
+        phi_n = np.concatenate(
+            [2 * np.pi * np.ones(phi_n.shape[1])[np.newaxis, ...], phi_n], axis=0
+        )
+        sines = np.sin(phi_n)
+        sines[0, :] = 1
+        cum_sines = np.cumprod(sines, axis=0)
+        cosines = np.roll(np.cos(phi_n), -1, axis=0)
+        return cum_sines * cosines * r
+
+
+def above_mu_min_angular_generator(bins, dims2avg, angular_resolution=0.1, mu=0.97):
+    r"""
+    Returns a set of spherical coordinates above a certain :math:`\\mu` value.
+
+    Parameters
+    ----------
+    bins : array-like
+        1D array of radii at which we want to spherically average the field.
+    dims2avg : int
+        The number of dimensions to average over.
+    angular_resolution : float, optional
+        The angular resolution in radians for the sample points for the interpolation.
+    mu : float, optional
+        The minimum value of :math:`\\cos(\theta), \theta = \arctan (k_\\perp/k_\\parallel)`
+        for the sample points generated for the interpolation.
+
+    Returns
+    -------
+    r_n : array-like
+        1D array of radii.
+    phi_n : array-like
+        2D array of azimuthal angles with shape (ndim-1, N), where N is the number of points.
+        phi_n[0,:] :math:`\\in [0,2*\\pi]`, and phi_n[1:,:] :math:`\\in [0,\\pi]`.
+    """
+
+    def generator():
+        r_n, phi_n = regular_angular_generator(
+            bins, dims2avg, angular_resolution=angular_resolution
+        )
+        # sine because the phi_n are wrt x-axis and we need them wrt z-axis.
+        if len(phi_n) == 1:
+            mask = np.sin(phi_n[0, :]) >= mu
+        else:
+            mask = np.all(np.sin(phi_n[1:, :]) >= mu, axis=0)
+        return r_n[mask], phi_n[:, mask]
+
+    return generator()
```

Comment:
You should be returning just `generator` not `generator()`. Also, `generator` must accept some arguments, because you call it with arguments in `_sample_coords_interpolate`.